crates/sargon/src/core/types/intent_discriminator.rs

```diff
@@ -0,0 +1,157 @@
+pub use crate::prelude::*;
+
+// Generate the FfiConverter needed by UniFFI for newtype `IntentDiscriminator`.
+uniffi::custom_newtype!(IntentDiscriminatorSecretMagic, u64);
+
+#[derive(
+    Clone, Copy, PartialEq, Eq, Hash, derive_more::Display, derive_more::Debug,
+)]
+pub struct IntentDiscriminatorSecretMagic(pub u64);
+
+impl IntentDiscriminatorSecretMagic {
+    /// Generates 8 random bytes and creates a IntentDiscriminator (u64) from it.
+    pub fn random() -> Self {
+        let value = u64::from_be_bytes(
+            generate_bytes::<8>()
+                .as_slice()
+                .try_into()
+                .expect("It is 8 bytes."),
+        );
+
+        Self(value)
+    }
+}
+
+/// A random number generated part of an intent header,
+/// ensuring every intent is unique even though its
+/// transaction manifest might be equal. This intent discriminator is
+/// generated by wallets for incoming intents.
+#[derive(
+    Clone,
+    Copy,
+    PartialEq,
+    Eq,
+    Hash,
+    derive_more::Display,
+    derive_more::Debug,
+    uniffi::Record,
+)]
+#[display("{}", self.secret_magic)]
+#[debug("{}", self.secret_magic)]
+pub struct IntentDiscriminator {
```

Comment:
maybe worth mentioning in doc that this is acting like a `nonce`, very similar to the nonce of the TransactionHeader, because it is, right?